Add ReorderedPartitionsRDD

tpoterba opened the pull request:
This allows us to coerce as sorted 'files.chr*.vcf',
no matter the order of the files in globbing.

cseed commented on src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala:
```diff
@@ -38,72 +44,80 @@ object OrderedRDD {
-      val sortedness = keyInfo.map(_.sortedness).min
+      val partitioner = OrderedPartitioner[PK, K](sortedKeyInfo.init.map(_.max))
+      val sortedness = sortedKeyInfo.map(_.sortedness).min
+      val scrambledRDD = new ReorderedPartitionsRDD[(K, V)](rdd, sortedKeyInfo.map(_.partIndex))
```
I'd changed `scrambled` to `reordered` or something.  Scrambled sounds like you're creating chaos, but you're creating order.

cseed commented on src/main/scala/org/broadinstitute/hail/sparkextras/ReorderedPartitionsRDD.scala:
```diff
@@ -0,0 +1,28 @@
+
+case class ReorderedPartition(index: Int, oldPartition: Partition) extends Partition
+
+class ReorderedPartitionsRDD[T](prev: RDD[T], oldIndices: Array[Int])(implicit tct: ClassTag[T])
```
`prev` should be `@transient` and you should implement clearDependencies.

cseed commented on src/test/scala/org/broadinstitute/hail/utils/OrderedRDDSuite.scala:
```diff
@@ -75,15 +75,15 @@ class OrderedRDDSuite extends SparkSuite {
 
     property("fullySorted") = Prop.forAll(sorted) { case (nPar, s) =>
-      check(sc.parallelize(s, nPar).toOrderedRDD[Locus])
+      check(sc.parallelize(s, nPar).toOrderedRDD)
```
Should you call coerce and verify the status for `sorted` and `locusSorted`?  Might also have an empty multi-partition one.

cseed commented on src/test/scala/org/broadinstitute/hail/utils/OrderedRDDSuite.scala:
```diff
@@ -151,6 +151,22 @@ class OrderedRDDSuite extends SparkSuite {
         .forall { case (v1, v2) => v1 == v2 }
     }
+
+    property("partitionReordering") = Prop.forAll(sorted) { case (nPar, is) =>
```
Should you run this against `sorted`, `locusSorted` or `random`?

tpoterba replied:
`random` will end up doing a shuffle.  I'll choose from sorted or locusSorted.

cseed commented on src/test/scala/org/broadinstitute/hail/utils/OrderedRDDSuite.scala:
```diff
@@ -151,6 +151,22 @@ class OrderedRDDSuite extends SparkSuite {
+
+      val rdd = sc.parallelize(is, nPar)
+      val shuffledIndices = Array.range(0, nPar).sortBy(_ => scala.util.Random.nextDouble())
+      val scrambled = new ReorderedPartitionsRDD(rdd, shuffledIndices)
```
Should we add `RichRDD.reorderPartitions(oldIndices: Array[Int])`?

cseed commented on src/test/scala/org/broadinstitute/hail/utils/OrderedRDDSuite.scala:
```diff
@@ -151,6 +151,22 @@ class OrderedRDDSuite extends SparkSuite {
+
+
+      status == status2 && result.fullOuterJoin(result2)
+        .collect()
```
You can do `forall` directly against the RDDs.

cseed commented on src/main/scala/org/broadinstitute/hail/sparkextras/ReorderedPartitionsRDD.scala:
```diff
@@ -0,0 +1,28 @@
+
+import scala.reflect.ClassTag
+
+case class ReorderedPartition(index: Int, oldPartition: Partition) extends Partition
```
`ReorderedPartitionsRDDPartition` is traditional.

cseed commented on src/test/scala/org/broadinstitute/hail/utils/OrderedRDDSuite.scala:
```diff
@@ -151,6 +151,22 @@ class OrderedRDDSuite extends SparkSuite {
+    property("partitionReordering") = Prop.forAll(sorted) { case (nPar, is) =>
+
+      val rdd = sc.parallelize(is, nPar)
+      val shuffledIndices = Array.range(0, nPar).sortBy(_ => scala.util.Random.nextDouble())
```
This is a annoying change since the Prop/Gen syntax is so bad, but it would be natural to have `Gen.shuffle(Gen.const(0 until nPar))` as another generator item.

cseed commented:
Back to you.

danking commented on src/main/scala/org/broadinstitute/hail/sparkextras/BlockedRDD.scala:
```diff
@@ -0,0 +1,68 @@
+      return Seq.empty
+
+    val m = locationAvail.values.max
+    locationAvail.filter(_._2 == m)
```
I feel like the standard library should (but seemingly doesn't) have a

```scala
trait Seq[+A] {
  def maxesBy[C[_], B](f: (A) => B)(implicit cbf: CanBuildFrom[Nothing, B, C[B]]): C[B] = {
    var c = cbf()
    if (this.isEmpty)
      c.result
    else {
      var biggest = this.head
      c += biggest
      for (a <- this.tail) {
        val b = f(a)
        if (b > biggest) {
          c = cbf()
          biggest = b
          c += biggest
        } else if (b == biggest) {
          c += b
        }
      }
      c.result
    }
  }
}
```

cseed replied:
I'm good wth this.  You can call `c.clear()` instead of constructing another builder at `c = cbf()`.

cseed replied:
Did you not want to throw this in?

danking commented on src/main/scala/org/broadinstitute/hail/sparkextras/BlockedRDD.scala:
```diff
@@ -0,0 +1,68 @@
+    val locationAvail = range.flatMap(i =>
+      prev.preferredLocations(prevPartitions(i)))
+      .groupBy(identity)
+      .mapValues(_.length)
```
would
```scala
range.flatMap(i => prev.preferredLocations(prevPartitions(i)).map((_, _.length)))
  .toMap
```
be more explicit about what's going on? Not sure if the performance is any different.

cseed replied:
Two comments:

 - I don't think that's the same.  I want to count how many previous partitions prefer each location.  Hence the `groupBy`.

 - Syntactic note, each instance of `_` is a separate argument.  If you want to use an argument multiple times, you can't use the implicit lambda syntax.

danking commented on src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala:
```diff
@@ -38,72 +44,80 @@ object OrderedRDD {
-        })
+    val keyInfo = keys.mapPartitionsWithIndex { case (i, it) =>
+      if (it.hasNext)
+        Iterator(PartitionKeyInfo.apply(i, it))
```
unnecessary `.apply`, I think?

danking commented on src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala:
```diff
@@ -2,20 +2,26 @@ package org.broadinstitute.hail.sparkextras
 
 object OrderedRDD {
 
+  final val ORDERED_PARTITIONER = 0
```
might be worth a

```scala
type COERCION_METHOD = Int
```
too

danking commented on src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala:
```diff
@@ -38,72 +44,80 @@ object OrderedRDD {
-      val partitioner = OrderedPartitioner(keyInfo.init.map(pi => pi.max))
-      val sortedness = keyInfo.map(_.sortedness).min
+      val partitioner = OrderedPartitioner[PK, K](sortedKeyInfo.init.map(_.max))
+      val sortedness = sortedKeyInfo.map(_.sortedness).min
```
Is it usually the case that every partition has the same sortedness? In particular I'm wondering if you ever have one unsorted partition and the rest sorted.

tpoterba replied:
This isn't something we can assume.  In particular, if the dataset is locus-sorted but not variant-sorted, we would expect some partitions to be K-sorted and some to be T-sorted.

danking commented on src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala:
```diff
@@ -38,72 +44,80 @@ object OrderedRDD {
 
     if (rdd.partitions.isEmpty)
-      return empty(rdd.sparkContext)
+      return (ORDERED_PARTITIONER, empty(rdd.sparkContext))
```
Is this `AS_IS`?

tpoterba replied:
I'm not sure.  It doesn't really matter, I guess.  I think I used AS_IS to indicate that you spent an action to check, but then you're done  The ORDERED_PARTITIONER really means that no work was done, which happens when the partitioner is ordered or the RDD is empty

danking commented on src/test/scala/org/broadinstitute/hail/utils/OrderedRDDSuite.scala:
```diff
@@ -151,10 +158,29 @@ class OrderedRDDSuite extends SparkSuite {
+      val (status2, result2) = OrderedRDD.coerce(scrambled)
+
+      status == status2 && result.fullOuterJoin(result2)
+        .forall { case (v, (s1, s2)) => s1 == s2 }
```
This makes me wish the quick check library had a way to point out what lead to the falsification of the property. Here we really have a huge number of conjunctions, and it might not be obvious which one is causing the problem.

cseed replied:
I agree, although I have no idea what to do about this.

danking commented on src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala:
```diff
@@ -38,72 +44,80 @@ object OrderedRDD {
+  }
+
+  def coerce[PK, K, V](rdd: RDD[(K, V)], fastKeys: Option[RDD[K]] = None)
+    (implicit kOk: OrderedKey[PK, K]): (Int, OrderedRDD[PK, K, V]) = {
```
It seems a lot of places toss out the first element of the pair, I wonder if there should be a separate method for those places that does the throwing away itself? Keeps the API slimmer.

tpoterba replied:
That's what the `apply` above is for.  This lets us test freely (and maybe even helps when the query optimizer comes to town) but the previous API doesn't change.

danking replied:
Ah yep, totally missed this.

tpoterba commented:
Addressed comments, ready for another look.  Change to `Gen.partition` revealed a few other bugs in master, which don't have elegant solutions:

1.  orderedPartitioner either needs an `empty` boolean flag, or a special `EmptyOrderedPartitioner` where `partitioner.nPartitions` is 0.  Otherwise, rdd.fullOuterJoin(other) crashes because it looks at partitioner nPartitions instead of rdd nPartitions.  This fix involves more logic in VSM.read as well.

2.  OrderedLeftJoinRDD needs to cache partitions and pass them to OrderedRDDIterator.  There is no guarantee that `getPartitions` can be called on an executor, and this was crashing in particular for ParallelCollectionRDDs.

danking commented on src/main/scala/org/broadinstitute/hail/sparkextras/OrderedPartitioner.scala:
```diff
@@ -13,22 +13,23 @@ import scala.reflect.{ClassTag, classTag}
 
-  require(rangeBounds.isEmpty ||
-    rangeBounds.zip(rangeBounds.tail).forall { case (left, right) => left < right })
+  require(!empty || rangeBounds.isEmpty)
```
might be clearer with implication
```scala
import org.broadinstitute.hail.Utils.richBoolean
// ...
require(empty ==> rangeBounds.isEmpty)
```

cseed replied:
Agreed, but I think by default that's only available in the test framework.  Or did we add it somewhere?

```
scala> true ==> false
<console>:8: error: value ==> is not a member of Boolean
              true ==> false
                   ^
```

danking commented:
I suppose the more "object oriented" way would be to have two classes. I don't have strong feelings either way, I think it's fine as it is.

cseed commented on src/main/scala/org/broadinstitute/hail/sparkextras/OrderedPartitioner.scala:
```diff
@@ -13,22 +13,23 @@ import scala.reflect.{ClassTag, classTag}
 import scala.util.hashing.{MurmurHash3, byteswap32}
 
-case class OrderedPartitioner[PK, K](rangeBounds: Array[PK], ascending: Boolean = true)(implicit val kOk: OrderedKey[PK, K])
+case class OrderedPartitioner[PK, K](rangeBounds: Array[PK], ascending: Boolean = true, empty: Boolean = false)
```
I'm worried about this `empty` business.  It seems error prone and I'm not convinced it is getting set correctly everywhere, e.g. `OrderedRDD.apply(rdd, ranges)`.  I think it should take a non-default `numPartitions: Int` argument.

cseed commented on src/main/scala/org/broadinstitute/hail/sparkextras/OrderedLeftJoinRDD.scala:
```diff
@@ -75,6 +76,8 @@ class OrderedLeftJoinRDD[PK, K, V1, V2](left: OrderedRDD[PK, K, V1], right: Orde
   extends RDD[(K, (V1, Option[V2]))](left.sparkContext, Seq(new OneToOneDependency(left),
     new OrderedDependency(left.orderedPartitioner, right.orderedPartitioner, right)): Seq[Dependency[_]]) {
 
+  val cachedRightPartitions = right.partitions
```
I'd just call this `rightPartitions` (and maybe make it private?)

cseed commented on src/main/scala/org/broadinstitute/hail/variant/VariantSampleMatrix.scala:
```diff
@@ -162,34 +163,30 @@ object VariantSampleMatrix {
-    val orderedRDD = partitioner match {
-      case Some(p) =>
-        OrderedRDD[Locus, Variant, (Annotation, Iterable[Genotype])](readRDD(skipGenotypes), p)
+    val partitioner = try {
```
The indentation is wrong here.

cseed commented on src/main/scala/org/broadinstitute/hail/variant/VariantSampleMatrix.scala:
```diff
@@ -706,17 +703,18 @@ class VariantSampleMatrix[T](val metadata: VariantMetadata,
-        }), gs)
-      }.asOrderedRDD[Locus])
+      rdd = rdd.mapValuesWithKey {
+        case (v, (va, gs)) =>
```
Unnecessary change in indentation (I think?).  I've been seeing IntelliJ do this sporadically and I'm not sure what's up with that.

cseed commented on src/main/scala/org/broadinstitute/hail/variant/VariantSampleMatrix.scala:
```diff
@@ -1008,9 +1006,10 @@ class RichVDS(vds: VariantDataset) {
-          (f2(f1(va)), gs.lazyMap(g => g.copy(fakeRef = false)))
-        }.asOrderedRDD[Locus])
+        rdd = vds1.rdd.mapValuesWithKey {
+          case (v, (va, gs)) =>
```
Again.

cseed commented on src/test/scala/org/broadinstitute/hail/variant/vsm/VSMSuite.scala:
```diff
@@ -170,7 +170,7 @@ class VSMSuite extends SparkSuite {
 
   @Test def testReadWrite() {
-    val p = forAll(VariantSampleMatrix.gen[Genotype](sc, VSMSubgen.random)) { (vsm: VariantSampleMatrix[Genotype]) =>
+    val p = forAll(VariantSampleMatrix.gen[Genotype](sc, VSMSubgen.random).filter(_.nVariants == 0)) { (vsm: VariantSampleMatrix[Genotype]) =>
```
`.filter(_.nVariants == 0)`?  Whoa, that can't be right.

tpoterba replied:
ah yup, used this to catch the empty RDD bug and forgot to change it back!

cseed commented on src/test/scala/org/broadinstitute/hail/variant/vsm/VSMSuite.scala:
```diff
@@ -306,7 +306,7 @@ class VSMSuite extends SparkSuite {
   @Test def testCoalesce() {
     val g = for (
-      vsm <- VariantSampleMatrix.gen[Genotype](sc, VSMSubgen.random);
+      vsm <- VariantSampleMatrix.gen[Genotype](sc, VSMSubgen.random).filter(_.nPartitions > 0);
```
We should be able to call `coalesce` on RDDs with empty partitions.  This needs to get fixed.

cseed commented:
OK, I think that's everything, back to 

I found this change hard to reason about.  Part of it was the layered changes.  Let's try to keep unrelated changes in unrelated commits.  Part of it is the inherent complexity of the Ordered stuff and Spark internals, although I don't have specific suggestions beyond my comments above.

tpoterba commented:
Addressed comments, back to you.

cseed commented on src/main/scala/org/broadinstitute/hail/sparkextras/OrderedLeftJoinRDD.scala:
```diff
@@ -9,22 +9,23 @@ import org.broadinstitute.hail.utils.richUtils.RichPairIterator
 
 class OrderedRDDIterator[PK, K, V](
   rdd: OrderedRDD[PK, K, V],
+  cachedPartitions: Array[Partition],
```
Did you not want to just rename this `partitions`?  Perhaps with a quick comment on why you can't just access `rdd.partitions`?

cseed commented on src/main/scala/org/broadinstitute/hail/sparkextras/OrderedPartitioner.scala:
```diff
@@ -13,22 +13,24 @@ import scala.reflect.{ClassTag, classTag}
 
-  require(rangeBounds.isEmpty ||
-    rangeBounds.zip(rangeBounds.tail).forall { case (left, right) => left < right })
+  require(nPartitions == 0 || nPartitions == rangeBounds.length + 1,
```
Shouldn't the first condition be `(nPartitions == 0 && rangeBounds.isEmpty)`?

cseed commented on src/main/scala/org/broadinstitute/hail/sparkextras/OrderedPartitioner.scala:
```diff
@@ -83,18 +85,18 @@ case class OrderedPartitioner[PK, K](rangeBounds: Array[PK], ascending: Boolean
     val ascending = in.readBoolean()
     val rangeBounds = in.readObject().asInstanceOf[Array[PK]]
-    OrderedPartitioner(rangeBounds, ascending)
+    OrderedPartitioner(rangeBounds, rangeBounds.length + 1, ascending)
```
Isn't this the same as the original `def nPartitions = rangeBounds.length + 1` bug?  You should read in the RDD in `VSM.read` and then pass in the number of partitions.

cseed commented on src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala:
```diff
@@ -42,82 +47,87 @@ object OrderedRDD {
       info("Ordering unsorted dataset with network shuffle")
       val ranges = calculateKeyRanges(keys.map(kOk.project))
-      apply(rdd, ranges)
+      (SHUFFLE, shuffle(rdd, OrderedPartitioner(ranges, ranges.length + 1)))
```
Shouldn't this be `rdd.partitions.length`?

cseed commented on src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala:
```diff
@@ -391,11 +338,11 @@ class OrderedRDD[PK, K, V] private(rdd: RDD[(K, V)], val orderedPartitioner: Ord
+      warn(s"coalesced to ${ newPartEnd.length } partitions, less than requested $maxPartitions")
 
-    new OrderedRDD[PK, K, V](new BlockedRDD(persisted, newPartEnd), OrderedPartitioner(newRangeBounds))
+    val newRangeBounds = Array.tabulate(newPartEnd.length - 1) { i => orderedPartitioner.rangeBounds(newPartEnd(i)) }
```
Does this do the same thing?  It doesn't seem like an improvement.

cseed commented:
OK, I'm done.

tpoterba commented:
Back to you!
